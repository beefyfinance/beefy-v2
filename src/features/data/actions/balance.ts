import { createAsyncThunk } from '@reduxjs/toolkit';
import type { BeefyState } from '../../../redux-types';
import type { FetchAllBalancesResult } from '../apis/balance/balance-types';
import { getBalanceApi } from '../apis/instances';
import type { ChainEntity } from '../entities/chain';
import { selectAllTokenWhereUserCouldHaveBalance } from '../selectors/balance';
import { selectBoostById, selectBoostsByChainId } from '../selectors/boosts';
import { selectChainById } from '../selectors/chains';
import { selectTokenByAddress } from '../selectors/tokens';
import { selectAllGovVaultsByChainId } from '../selectors/vaults';
import { selectWalletAddress } from '../selectors/wallet';
import type { TokenEntity } from '../entities/token';
import type { VaultEntity } from '../entities/vault';
import { isGovVault } from '../entities/vault';
import { uniqueTokens } from '../../../helpers/tokens';

export interface FetchAllBalanceActionParams {
  chainId: ChainEntity['id'];
  walletAddress: string;
}

export interface FetchAllBalanceFulfilledPayload {
  chainId: ChainEntity['id'];
  walletAddress: string;
  data: FetchAllBalancesResult;
  // reducers need the state (balance)
  state: BeefyState;
}

export const fetchAllBalanceAction = createAsyncThunk<
  FetchAllBalanceFulfilledPayload,
  FetchAllBalanceActionParams,
  { state: BeefyState }
>('balance/fetchAllBalanceAction', async ({ chainId, walletAddress }, { getState }) => {
  const state = getState();

  const userAddress = walletAddress ?? selectWalletAddress(state);
  const chain = selectChainById(state, chainId);
  const api = await getBalanceApi(chain);

  const tokens = selectAllTokenWhereUserCouldHaveBalance(state, chainId).map(address =>
    selectTokenByAddress(state, chain.id, address)
  );
  // maybe have a way to retrieve those easily
  const boosts = selectBoostsByChainId(state, chainId).map(boostId =>
    selectBoostById(state, boostId)
  );
  const govVaults = selectAllGovVaultsByChainId(state, chain.id);

  const data = await api.fetchAllBalances(getState(), tokens, govVaults, boosts, userAddress);
  return {
    chainId,
    walletAddress: userAddress,
    data,
    state: getState(),
  };
});

export type FetchBalanceParams = {
  chainId: ChainEntity['id'];
  tokens?: TokenEntity[];
  vaults?: VaultEntity[];
};

export const fetchBalanceAction = createAsyncThunk<
  FetchAllBalanceFulfilledPayload,
  FetchBalanceParams,
  { state: BeefyState }
>(
  'balance/fetchBalanceAction',
  async ({ chainId, tokens: requestedTokens = [], vaults = [] }, { getState }) => {
    const state = getState();

    const walletAddress = selectWalletAddress(state);
    const chain = selectChainById(state, chainId);
    const api = await getBalanceApi(chain);

    const tokens = requestedTokens;
    const govVaults = [];
    const boosts = [];

    if (vaults.length) {
      for (const vault of vaults) {
        if (isGovVault(vault)) {
          govVaults.push(vault);
        } else {
          tokens.push(selectTokenByAddress(state, chain.id, vault.depositTokenAddress));
          tokens.push(selectTokenByAddress(state, chain.id, vault.earnedTokenAddress));
        }
      }
    }

    const data = await api.fetchAllBalances(
      getState(),
      uniqueTokens(tokens),
      govVaults,
      boosts,
      walletAddress
    );

    return {
      chainId,
      walletAddress,
      data,
      state: getState(),
    };
  }
);
