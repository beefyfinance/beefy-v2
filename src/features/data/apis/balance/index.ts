import { BalanceAPI } from './balance';
import { BalanceMcV2API } from './balance-multicallv2';

export { BalanceAPI, BalanceMcV2API };
