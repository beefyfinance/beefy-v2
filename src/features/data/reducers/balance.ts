import { createSlice } from '@reduxjs/toolkit';
import BigNumber from 'bignumber.js';
import { byDecimals } from '../../../helpers/format';
import { fetchAllBalanceAction } from '../actions/balance';
import { BoostEntity } from '../entities/boost';
import { ChainEntity } from '../entities/chain';
import { TokenEntity } from '../entities/token';
import { VaultEntity } from '../entities/vault';
import { selectBoostById } from '../selectors/boosts';
import { selectTokenById } from '../selectors/tokens';
import {
  selectIsStandardVaultEarnTokenId,
  selectVaultByEarnTokenId,
  selectVaultById,
  selectVaultPricePerFullShare,
} from '../selectors/vaults';
import { accountHasChanged, walletHasDisconnected } from './wallet';

/**
 * State containing user balances state
 */
export interface BalanceState {
  depositedVaultIds: {
    govVaults: { chainId: ChainEntity['id']; vaultId: VaultEntity['id'] }[];
    standardVaults: { chainId: ChainEntity['id']; vaultId: VaultEntity['id'] }[];
  };
  // computed total balance accounting for boost
  // mostly useful for the vault list
  deposited: {
    [vaultId: VaultEntity['id']]: { balance: BigNumber; shares: BigNumber };
  };

  byChainId: {
    [chainId: ChainEntity['id']]: {
      byTokenId: {
        [tokenId: TokenEntity['id']]: {
          balance: BigNumber;
        };
      };
      byBoostId: {
        [boostId: BoostEntity['id']]: {
          balance: BigNumber;
          rewards: BigNumber;
        };
      };
      byGovVaultId: {
        [vaultId: VaultEntity['id']]: {
          balance: BigNumber;
          rewards: BigNumber;
        };
      };
    };
  };
}
export const initialBalanceState: BalanceState = {
  byChainId: {},
  deposited: {},
  depositedVaultIds: { govVaults: [], standardVaults: [] },
};

export const balanceSlice = createSlice({
  name: 'balance',
  initialState: initialBalanceState,
  reducers: {
    // standard reducer logic, with auto-generated action types per reducer
  },
  extraReducers: builder => {
    builder.addCase(accountHasChanged, sliceState => {
      sliceState.byChainId = {};
      sliceState.depositedVaultIds = { govVaults: [], standardVaults: [] };
    });
    builder.addCase(walletHasDisconnected, sliceState => {
      sliceState.byChainId = {};
      sliceState.depositedVaultIds = { govVaults: [], standardVaults: [] };
    });

    builder.addCase(fetchAllBalanceAction.fulfilled, (sliceState, action) => {
      const state = action.payload.state;
      const chainId = action.payload.chainId;

      /**
       * Ingest token data
       */
      for (const tokenBalance of action.payload.data.tokens) {
        if (sliceState.byChainId[chainId] === undefined) {
          sliceState.byChainId[chainId] = { byTokenId: {}, byBoostId: {}, byGovVaultId: {} };
        }

        // only update data if necessary
        const stateForToken = sliceState.byChainId[chainId].byTokenId[tokenBalance.tokenId];
        if (
          // only if balance is non-zero
          !tokenBalance.amount.isZero() &&
          // and state isn't already there
          // if it's there, only if amount differ
          (stateForToken === undefined || !stateForToken.balance.isEqualTo(tokenBalance.amount))
        ) {
          sliceState.byChainId[chainId].byTokenId[tokenBalance.tokenId] = {
            balance: tokenBalance.amount,
          };

          // now, if this is a vault token, we want to mark this vault as deposited
          if (selectIsStandardVaultEarnTokenId(state, chainId, tokenBalance.tokenId)) {
            const vaultId = selectVaultByEarnTokenId(state, chainId, tokenBalance.tokenId);

            sliceState.depositedVaultIds.standardVaults.push({ chainId, vaultId });

            // add to total balance
            const vault = selectVaultById(state, vaultId);
            const oracleToken = selectTokenById(state, vault.chainId, vault.oracleId);
            const ppfs = selectVaultPricePerFullShare(state, vault.id);

            if (!sliceState.deposited[vaultId]) {
              sliceState.deposited[vaultId] = {
                balance: new BigNumber(0),
                shares: new BigNumber(0),
              };
            }
            const balanceSingle = byDecimals(
              new BigNumber(tokenBalance.amount).multipliedBy(byDecimals(ppfs)).toFixed(8),
              oracleToken.decimals
            );
            const vaultTotalState = sliceState.deposited[vaultId];
            vaultTotalState.balance = vaultTotalState.balance.plus(balanceSingle);
            vaultTotalState.shares = vaultTotalState.shares.plus(tokenBalance.amount);
          }
        }
      }

      /**
       * Ingest boost data
       */
      for (const boostBalance of action.payload.data.boosts) {
        if (sliceState.byChainId[chainId] === undefined) {
          sliceState.byChainId[chainId] = { byTokenId: {}, byBoostId: {}, byGovVaultId: {} };
        }

        // only update data if necessary
        const stateForBoost = sliceState.byChainId[chainId].byBoostId[boostBalance.boostId];
        if (
          stateForBoost === undefined ||
          !stateForBoost.balance.isEqualTo(boostBalance.balance) ||
          !stateForBoost.rewards.isEqualTo(boostBalance.rewards)
        ) {
          sliceState.byChainId[chainId].byBoostId[boostBalance.boostId] = {
            balance: boostBalance.balance,
            rewards: boostBalance.rewards,
          };
        }

        // add to total balance of the target vault
        const boost = selectBoostById(state, boostBalance.boostId);
        const vault = selectVaultById(state, boost.vaultId);
        const oracleToken = selectTokenById(state, vault.chainId, vault.oracleId);
        const ppfs = selectVaultPricePerFullShare(state, vault.id);
        if (!sliceState.deposited[vault.id]) {
          sliceState.deposited[vault.id] = {
            balance: new BigNumber(0),
            shares: new BigNumber(0),
          };
        }
        const balanceSingle = byDecimals(
          boostBalance.balance.multipliedBy(byDecimals(ppfs)),
          oracleToken.decimals
        );
        const vaultTotalState = sliceState.deposited[vault.id];
        vaultTotalState.balance = vaultTotalState.balance.plus(balanceSingle);
        vaultTotalState.shares = vaultTotalState.shares.plus(boostBalance.balance);
      }

      /**
       * Ingest gov vaults data
       */
      for (const vaultBalance of action.payload.data.govVaults) {
        const vaultId = vaultBalance.vaultId;
        if (sliceState.byChainId[chainId] === undefined) {
          sliceState.byChainId[chainId] = { byTokenId: {}, byBoostId: {}, byGovVaultId: {} };
        }

        // only update data if necessary
        const stateForVault = sliceState.byChainId[chainId].byGovVaultId[vaultId];
        if (
          // only if balance is non-zero
          !vaultBalance.rewards.isZero() &&
          // and state isn't already there
          // if it's there, only if amount differ
          (stateForVault === undefined || !stateForVault.rewards.isEqualTo(vaultBalance.rewards))
        ) {
          const vaultState = {
            rewards: vaultBalance.rewards,
            balance: vaultBalance.balance,
          };
          sliceState.byChainId[chainId].byGovVaultId[vaultId] = vaultState;
          sliceState.depositedVaultIds.govVaults.push({ chainId, vaultId });

          // add to total balance
          const vault = selectVaultById(state, vaultId);
          const oracleToken = selectTokenById(state, vault.chainId, vault.oracleId);

          if (!sliceState.deposited[vaultId]) {
            sliceState.deposited[vaultId] = {
              balance: new BigNumber(0),
              shares: new BigNumber(0),
            };
          }
          const balanceSingle = byDecimals(vaultBalance.balance, oracleToken.decimals);
          const vaultTotalState = sliceState.deposited[vaultId];
          vaultTotalState.balance = vaultTotalState.balance.plus(balanceSingle);
          vaultTotalState.shares = vaultTotalState.shares.plus(vaultBalance.balance);
        }
      }
    });
  },
});
