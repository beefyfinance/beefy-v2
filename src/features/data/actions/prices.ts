import { createAsyncThunk } from '@reduxjs/toolkit';
import { BeefyAPITokenPricesResponse } from '../apis/beefy';
import { getBeefyApi } from '../apis/instances';

export const fetchAllPricesAction = createAsyncThunk<BeefyAPITokenPricesResponse, {}>(
  'prices/fetchAllPricesAction',
  async () => {
    const api = getBeefyApi();
    const [prices, lpPrices] = await Promise.all([api.getPrices(), api.getLPs()]);
    return {
      ...prices,
      ...lpPrices,
    };
  }
);
