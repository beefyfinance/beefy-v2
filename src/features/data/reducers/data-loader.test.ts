import { fetchAllBalanceAction } from '../actions/balance';
import { fetchAllPricesAction } from '../actions/prices';
import { dataLoaderSlice, initialDataLoaderState } from './data-loader';

describe('Data Loader slice tests', () => {
  it('should update state on fulfilled global action', () => {
    let sliceState = initialDataLoaderState;

    expect(sliceState.global.prices.alreadyLoadedOnce).toBe(false);

    sliceState = dataLoaderSlice.reducer(sliceState, { type: fetchAllPricesAction.pending });
    expect(sliceState).toMatchSnapshot();

    expect(sliceState.global.prices.alreadyLoadedOnce).toBe(false);

    sliceState = dataLoaderSlice.reducer(sliceState, { type: fetchAllPricesAction.fulfilled });
    expect(sliceState).toMatchSnapshot();

    expect(sliceState.global.prices.alreadyLoadedOnce).toBe(true);

    sliceState = dataLoaderSlice.reducer(sliceState, {
      type: fetchAllPricesAction.rejected,
      error: 'fatality',
    });
    expect(sliceState).toMatchSnapshot();
    sliceState = dataLoaderSlice.reducer(sliceState, {
      type: fetchAllPricesAction.rejected,
      error: { message: 'fatality message' },
    });
    expect(sliceState).toMatchSnapshot();

    sliceState = dataLoaderSlice.reducer(sliceState, {
      type: fetchAllPricesAction.rejected,
      error: { name: 'fatality name' },
    });
    expect(sliceState).toMatchSnapshot();

    sliceState = dataLoaderSlice.reducer(sliceState, {
      type: fetchAllPricesAction.rejected,
      error: { code: 1234 },
    });
    expect(sliceState).toMatchSnapshot();

    expect(sliceState.global.prices.alreadyLoadedOnce).toBe(true);

    sliceState = dataLoaderSlice.reducer(sliceState, { type: fetchAllPricesAction.pending });
    expect(sliceState).toMatchSnapshot();

    expect(sliceState.global.prices.alreadyLoadedOnce).toBe(true);
  });

  it('should update state on fulfilled action by chain id', () => {
    let sliceState = initialDataLoaderState;

    expect(sliceState.byChainId['test']).toBe(undefined);
    sliceState = dataLoaderSlice.reducer(sliceState, {
      type: fetchAllBalanceAction.pending,
      meta: { arg: { chainId: 'test' } },
    });
    expect(sliceState).toMatchSnapshot();

    expect(sliceState.byChainId['test'].balance.alreadyLoadedOnce).toBe(false);

    sliceState = dataLoaderSlice.reducer(sliceState, {
      type: fetchAllBalanceAction.fulfilled,
      meta: { arg: { chainId: 'test' } },
    });
    expect(sliceState).toMatchSnapshot();

    expect(sliceState.byChainId['test'].balance.alreadyLoadedOnce).toBe(true);

    sliceState = dataLoaderSlice.reducer(sliceState, {
      type: fetchAllBalanceAction.rejected,
      error: 'fatality',
      meta: { arg: { chainId: 'test' } },
    });
    expect(sliceState).toMatchSnapshot();

    expect(sliceState.byChainId['test'].balance.alreadyLoadedOnce).toBe(true);

    sliceState = dataLoaderSlice.reducer(sliceState, {
      type: fetchAllBalanceAction.pending,
      meta: { arg: { chainId: 'test' } },
    });
    expect(sliceState).toMatchSnapshot();

    expect(sliceState.byChainId['test'].balance.alreadyLoadedOnce).toBe(true);
  });
});
