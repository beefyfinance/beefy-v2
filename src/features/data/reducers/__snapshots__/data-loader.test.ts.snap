// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Data Loader slice tests should update state on fulfilled action by chain id 1`] = `
Object {
  "byChainId": Object {
    "test": Object {
      "addressBook": Object {
        "alreadyLoadedOnce": false,
        "error": null,
        "status": "init",
      },
      "allowance": Object {
        "alreadyLoadedOnce": false,
        "error": null,
        "status": "init",
      },
      "balance": Object {
        "alreadyLoadedOnce": false,
        "error": null,
        "status": "pending",
      },
      "contractData": Object {
        "alreadyLoadedOnce": false,
        "error": null,
        "status": "init",
      },
    },
  },
  "global": Object {
    "addressBook": Object {
      "alreadyLoadedOnce": false,
      "error": null,
      "status": "init",
    },
    "apy": Object {
      "alreadyLoadedOnce": false,
      "error": null,
      "status": "init",
    },
    "boostForm": Object {
      "alreadyLoadedOnce": false,
      "error": null,
      "status": "init",
    },
    "boosts": Object {
      "alreadyLoadedOnce": false,
      "error": null,
      "status": "init",
    },
    "chainConfig": Object {
      "alreadyLoadedOnce": false,
      "error": null,
      "status": "init",
    },
    "depositForm": Object {
      "alreadyLoadedOnce": false,
      "error": null,
      "status": "init",
    },
    "prices": Object {
      "alreadyLoadedOnce": false,
      "error": null,
      "status": "init",
    },
    "vaults": Object {
      "alreadyLoadedOnce": false,
      "error": null,
      "status": "init",
    },
    "wallet": Object {
      "alreadyLoadedOnce": false,
      "error": null,
      "status": "init",
    },
    "withdrawForm": Object {
      "alreadyLoadedOnce": false,
      "error": null,
      "status": "init",
    },
    "zaps": Object {
      "alreadyLoadedOnce": false,
      "error": null,
      "status": "init",
    },
  },
  "instances": Object {
    "wallet": false,
  },
  "statusIndicator": Object {
    "open": false,
  },
}
`;

exports[`Data Loader slice tests should update state on fulfilled action by chain id 2`] = `
Object {
  "byChainId": Object {
    "test": Object {
      "addressBook": Object {
        "alreadyLoadedOnce": false,
        "error": null,
        "status": "init",
      },
      "allowance": Object {
        "alreadyLoadedOnce": false,
        "error": null,
        "status": "init",
      },
      "balance": Object {
        "alreadyLoadedOnce": true,
        "error": null,
        "status": "fulfilled",
      },
      "contractData": Object {
        "alreadyLoadedOnce": false,
        "error": null,
        "status": "init",
      },
    },
  },
  "global": Object {
    "addressBook": Object {
      "alreadyLoadedOnce": false,
      "error": null,
      "status": "init",
    },
    "apy": Object {
      "alreadyLoadedOnce": false,
      "error": null,
      "status": "init",
    },
    "boostForm": Object {
      "alreadyLoadedOnce": false,
      "error": null,
      "status": "init",
    },
    "boosts": Object {
      "alreadyLoadedOnce": false,
      "error": null,
      "status": "init",
    },
    "chainConfig": Object {
      "alreadyLoadedOnce": false,
      "error": null,
      "status": "init",
    },
    "depositForm": Object {
      "alreadyLoadedOnce": false,
      "error": null,
      "status": "init",
    },
    "prices": Object {
      "alreadyLoadedOnce": false,
      "error": null,
      "status": "init",
    },
    "vaults": Object {
      "alreadyLoadedOnce": false,
      "error": null,
      "status": "init",
    },
    "wallet": Object {
      "alreadyLoadedOnce": false,
      "error": null,
      "status": "init",
    },
    "withdrawForm": Object {
      "alreadyLoadedOnce": false,
      "error": null,
      "status": "init",
    },
    "zaps": Object {
      "alreadyLoadedOnce": false,
      "error": null,
      "status": "init",
    },
  },
  "instances": Object {
    "wallet": false,
  },
  "statusIndicator": Object {
    "open": false,
  },
}
`;

exports[`Data Loader slice tests should update state on fulfilled action by chain id 3`] = `
Object {
  "byChainId": Object {
    "test": Object {
      "addressBook": Object {
        "alreadyLoadedOnce": false,
        "error": null,
        "status": "init",
      },
      "allowance": Object {
        "alreadyLoadedOnce": false,
        "error": null,
        "status": "init",
      },
      "balance": Object {
        "alreadyLoadedOnce": true,
        "error": "fatality",
        "status": "rejected",
      },
      "contractData": Object {
        "alreadyLoadedOnce": false,
        "error": null,
        "status": "init",
      },
    },
  },
  "global": Object {
    "addressBook": Object {
      "alreadyLoadedOnce": false,
      "error": null,
      "status": "init",
    },
    "apy": Object {
      "alreadyLoadedOnce": false,
      "error": null,
      "status": "init",
    },
    "boostForm": Object {
      "alreadyLoadedOnce": false,
      "error": null,
      "status": "init",
    },
    "boosts": Object {
      "alreadyLoadedOnce": false,
      "error": null,
      "status": "init",
    },
    "chainConfig": Object {
      "alreadyLoadedOnce": false,
      "error": null,
      "status": "init",
    },
    "depositForm": Object {
      "alreadyLoadedOnce": false,
      "error": null,
      "status": "init",
    },
    "prices": Object {
      "alreadyLoadedOnce": false,
      "error": null,
      "status": "init",
    },
    "vaults": Object {
      "alreadyLoadedOnce": false,
      "error": null,
      "status": "init",
    },
    "wallet": Object {
      "alreadyLoadedOnce": false,
      "error": null,
      "status": "init",
    },
    "withdrawForm": Object {
      "alreadyLoadedOnce": false,
      "error": null,
      "status": "init",
    },
    "zaps": Object {
      "alreadyLoadedOnce": false,
      "error": null,
      "status": "init",
    },
  },
  "instances": Object {
    "wallet": false,
  },
  "statusIndicator": Object {
    "open": true,
  },
}
`;

exports[`Data Loader slice tests should update state on fulfilled action by chain id 4`] = `
Object {
  "byChainId": Object {
    "test": Object {
      "addressBook": Object {
        "alreadyLoadedOnce": false,
        "error": null,
        "status": "init",
      },
      "allowance": Object {
        "alreadyLoadedOnce": false,
        "error": null,
        "status": "init",
      },
      "balance": Object {
        "alreadyLoadedOnce": true,
        "error": null,
        "status": "pending",
      },
      "contractData": Object {
        "alreadyLoadedOnce": false,
        "error": null,
        "status": "init",
      },
    },
  },
  "global": Object {
    "addressBook": Object {
      "alreadyLoadedOnce": false,
      "error": null,
      "status": "init",
    },
    "apy": Object {
      "alreadyLoadedOnce": false,
      "error": null,
      "status": "init",
    },
    "boostForm": Object {
      "alreadyLoadedOnce": false,
      "error": null,
      "status": "init",
    },
    "boosts": Object {
      "alreadyLoadedOnce": false,
      "error": null,
      "status": "init",
    },
    "chainConfig": Object {
      "alreadyLoadedOnce": false,
      "error": null,
      "status": "init",
    },
    "depositForm": Object {
      "alreadyLoadedOnce": false,
      "error": null,
      "status": "init",
    },
    "prices": Object {
      "alreadyLoadedOnce": false,
      "error": null,
      "status": "init",
    },
    "vaults": Object {
      "alreadyLoadedOnce": false,
      "error": null,
      "status": "init",
    },
    "wallet": Object {
      "alreadyLoadedOnce": false,
      "error": null,
      "status": "init",
    },
    "withdrawForm": Object {
      "alreadyLoadedOnce": false,
      "error": null,
      "status": "init",
    },
    "zaps": Object {
      "alreadyLoadedOnce": false,
      "error": null,
      "status": "init",
    },
  },
  "instances": Object {
    "wallet": false,
  },
  "statusIndicator": Object {
    "open": true,
  },
}
`;

exports[`Data Loader slice tests should update state on fulfilled global action 1`] = `
Object {
  "byChainId": Object {},
  "global": Object {
    "addressBook": Object {
      "alreadyLoadedOnce": false,
      "error": null,
      "status": "init",
    },
    "apy": Object {
      "alreadyLoadedOnce": false,
      "error": null,
      "status": "init",
    },
    "boostForm": Object {
      "alreadyLoadedOnce": false,
      "error": null,
      "status": "init",
    },
    "boosts": Object {
      "alreadyLoadedOnce": false,
      "error": null,
      "status": "init",
    },
    "chainConfig": Object {
      "alreadyLoadedOnce": false,
      "error": null,
      "status": "init",
    },
    "depositForm": Object {
      "alreadyLoadedOnce": false,
      "error": null,
      "status": "init",
    },
    "prices": Object {
      "alreadyLoadedOnce": false,
      "error": null,
      "status": "pending",
    },
    "vaults": Object {
      "alreadyLoadedOnce": false,
      "error": null,
      "status": "init",
    },
    "wallet": Object {
      "alreadyLoadedOnce": false,
      "error": null,
      "status": "init",
    },
    "withdrawForm": Object {
      "alreadyLoadedOnce": false,
      "error": null,
      "status": "init",
    },
    "zaps": Object {
      "alreadyLoadedOnce": false,
      "error": null,
      "status": "init",
    },
  },
  "instances": Object {
    "wallet": false,
  },
  "statusIndicator": Object {
    "open": false,
  },
}
`;

exports[`Data Loader slice tests should update state on fulfilled global action 2`] = `
Object {
  "byChainId": Object {},
  "global": Object {
    "addressBook": Object {
      "alreadyLoadedOnce": false,
      "error": null,
      "status": "init",
    },
    "apy": Object {
      "alreadyLoadedOnce": false,
      "error": null,
      "status": "init",
    },
    "boostForm": Object {
      "alreadyLoadedOnce": false,
      "error": null,
      "status": "init",
    },
    "boosts": Object {
      "alreadyLoadedOnce": false,
      "error": null,
      "status": "init",
    },
    "chainConfig": Object {
      "alreadyLoadedOnce": false,
      "error": null,
      "status": "init",
    },
    "depositForm": Object {
      "alreadyLoadedOnce": false,
      "error": null,
      "status": "init",
    },
    "prices": Object {
      "alreadyLoadedOnce": true,
      "error": null,
      "status": "fulfilled",
    },
    "vaults": Object {
      "alreadyLoadedOnce": false,
      "error": null,
      "status": "init",
    },
    "wallet": Object {
      "alreadyLoadedOnce": false,
      "error": null,
      "status": "init",
    },
    "withdrawForm": Object {
      "alreadyLoadedOnce": false,
      "error": null,
      "status": "init",
    },
    "zaps": Object {
      "alreadyLoadedOnce": false,
      "error": null,
      "status": "init",
    },
  },
  "instances": Object {
    "wallet": false,
  },
  "statusIndicator": Object {
    "open": false,
  },
}
`;

exports[`Data Loader slice tests should update state on fulfilled global action 3`] = `
Object {
  "byChainId": Object {},
  "global": Object {
    "addressBook": Object {
      "alreadyLoadedOnce": false,
      "error": null,
      "status": "init",
    },
    "apy": Object {
      "alreadyLoadedOnce": false,
      "error": null,
      "status": "init",
    },
    "boostForm": Object {
      "alreadyLoadedOnce": false,
      "error": null,
      "status": "init",
    },
    "boosts": Object {
      "alreadyLoadedOnce": false,
      "error": null,
      "status": "init",
    },
    "chainConfig": Object {
      "alreadyLoadedOnce": false,
      "error": null,
      "status": "init",
    },
    "depositForm": Object {
      "alreadyLoadedOnce": false,
      "error": null,
      "status": "init",
    },
    "prices": Object {
      "alreadyLoadedOnce": true,
      "error": "fatality",
      "status": "rejected",
    },
    "vaults": Object {
      "alreadyLoadedOnce": false,
      "error": null,
      "status": "init",
    },
    "wallet": Object {
      "alreadyLoadedOnce": false,
      "error": null,
      "status": "init",
    },
    "withdrawForm": Object {
      "alreadyLoadedOnce": false,
      "error": null,
      "status": "init",
    },
    "zaps": Object {
      "alreadyLoadedOnce": false,
      "error": null,
      "status": "init",
    },
  },
  "instances": Object {
    "wallet": false,
  },
  "statusIndicator": Object {
    "open": true,
  },
}
`;

exports[`Data Loader slice tests should update state on fulfilled global action 4`] = `
Object {
  "byChainId": Object {},
  "global": Object {
    "addressBook": Object {
      "alreadyLoadedOnce": false,
      "error": null,
      "status": "init",
    },
    "apy": Object {
      "alreadyLoadedOnce": false,
      "error": null,
      "status": "init",
    },
    "boostForm": Object {
      "alreadyLoadedOnce": false,
      "error": null,
      "status": "init",
    },
    "boosts": Object {
      "alreadyLoadedOnce": false,
      "error": null,
      "status": "init",
    },
    "chainConfig": Object {
      "alreadyLoadedOnce": false,
      "error": null,
      "status": "init",
    },
    "depositForm": Object {
      "alreadyLoadedOnce": false,
      "error": null,
      "status": "init",
    },
    "prices": Object {
      "alreadyLoadedOnce": true,
      "error": "fatality message",
      "status": "rejected",
    },
    "vaults": Object {
      "alreadyLoadedOnce": false,
      "error": null,
      "status": "init",
    },
    "wallet": Object {
      "alreadyLoadedOnce": false,
      "error": null,
      "status": "init",
    },
    "withdrawForm": Object {
      "alreadyLoadedOnce": false,
      "error": null,
      "status": "init",
    },
    "zaps": Object {
      "alreadyLoadedOnce": false,
      "error": null,
      "status": "init",
    },
  },
  "instances": Object {
    "wallet": false,
  },
  "statusIndicator": Object {
    "open": true,
  },
}
`;

exports[`Data Loader slice tests should update state on fulfilled global action 5`] = `
Object {
  "byChainId": Object {},
  "global": Object {
    "addressBook": Object {
      "alreadyLoadedOnce": false,
      "error": null,
      "status": "init",
    },
    "apy": Object {
      "alreadyLoadedOnce": false,
      "error": null,
      "status": "init",
    },
    "boostForm": Object {
      "alreadyLoadedOnce": false,
      "error": null,
      "status": "init",
    },
    "boosts": Object {
      "alreadyLoadedOnce": false,
      "error": null,
      "status": "init",
    },
    "chainConfig": Object {
      "alreadyLoadedOnce": false,
      "error": null,
      "status": "init",
    },
    "depositForm": Object {
      "alreadyLoadedOnce": false,
      "error": null,
      "status": "init",
    },
    "prices": Object {
      "alreadyLoadedOnce": true,
      "error": "fatality name",
      "status": "rejected",
    },
    "vaults": Object {
      "alreadyLoadedOnce": false,
      "error": null,
      "status": "init",
    },
    "wallet": Object {
      "alreadyLoadedOnce": false,
      "error": null,
      "status": "init",
    },
    "withdrawForm": Object {
      "alreadyLoadedOnce": false,
      "error": null,
      "status": "init",
    },
    "zaps": Object {
      "alreadyLoadedOnce": false,
      "error": null,
      "status": "init",
    },
  },
  "instances": Object {
    "wallet": false,
  },
  "statusIndicator": Object {
    "open": true,
  },
}
`;

exports[`Data Loader slice tests should update state on fulfilled global action 6`] = `
Object {
  "byChainId": Object {},
  "global": Object {
    "addressBook": Object {
      "alreadyLoadedOnce": false,
      "error": null,
      "status": "init",
    },
    "apy": Object {
      "alreadyLoadedOnce": false,
      "error": null,
      "status": "init",
    },
    "boostForm": Object {
      "alreadyLoadedOnce": false,
      "error": null,
      "status": "init",
    },
    "boosts": Object {
      "alreadyLoadedOnce": false,
      "error": null,
      "status": "init",
    },
    "chainConfig": Object {
      "alreadyLoadedOnce": false,
      "error": null,
      "status": "init",
    },
    "depositForm": Object {
      "alreadyLoadedOnce": false,
      "error": null,
      "status": "init",
    },
    "prices": Object {
      "alreadyLoadedOnce": true,
      "error": "1234",
      "status": "rejected",
    },
    "vaults": Object {
      "alreadyLoadedOnce": false,
      "error": null,
      "status": "init",
    },
    "wallet": Object {
      "alreadyLoadedOnce": false,
      "error": null,
      "status": "init",
    },
    "withdrawForm": Object {
      "alreadyLoadedOnce": false,
      "error": null,
      "status": "init",
    },
    "zaps": Object {
      "alreadyLoadedOnce": false,
      "error": null,
      "status": "init",
    },
  },
  "instances": Object {
    "wallet": false,
  },
  "statusIndicator": Object {
    "open": true,
  },
}
`;

exports[`Data Loader slice tests should update state on fulfilled global action 7`] = `
Object {
  "byChainId": Object {},
  "global": Object {
    "addressBook": Object {
      "alreadyLoadedOnce": false,
      "error": null,
      "status": "init",
    },
    "apy": Object {
      "alreadyLoadedOnce": false,
      "error": null,
      "status": "init",
    },
    "boostForm": Object {
      "alreadyLoadedOnce": false,
      "error": null,
      "status": "init",
    },
    "boosts": Object {
      "alreadyLoadedOnce": false,
      "error": null,
      "status": "init",
    },
    "chainConfig": Object {
      "alreadyLoadedOnce": false,
      "error": null,
      "status": "init",
    },
    "depositForm": Object {
      "alreadyLoadedOnce": false,
      "error": null,
      "status": "init",
    },
    "prices": Object {
      "alreadyLoadedOnce": true,
      "error": null,
      "status": "pending",
    },
    "vaults": Object {
      "alreadyLoadedOnce": false,
      "error": null,
      "status": "init",
    },
    "wallet": Object {
      "alreadyLoadedOnce": false,
      "error": null,
      "status": "init",
    },
    "withdrawForm": Object {
      "alreadyLoadedOnce": false,
      "error": null,
      "status": "init",
    },
    "zaps": Object {
      "alreadyLoadedOnce": false,
      "error": null,
      "status": "init",
    },
  },
  "instances": Object {
    "wallet": false,
  },
  "statusIndicator": Object {
    "open": true,
  },
}
`;
